### PR TITLE
[tmva][sofie] Fix Conv2D padding for channels_first data format

### LINKED ISSUE #21661: [tmva][sofie] Conv2D with data_format=channels_first and padding=same gives wrong output shape
### Check duplicate issues.

- [x] Checked for duplicates

### Description

When parsing a Keras Conv2D layer with `data_format='channels_first'` and `padding='same'`, the SOFIE Keras parser computes padding using wrong indices. It always uses indices 1 and 2 for height and width, which is correct for channels_last (NHWC) but wrong for channels_first (NCHW). For channels_first the shape is (N, C, H, W) so H is at index 2 and W is at index 3. Because of this the output tensor shape does not match Keras and the inference results are wrong.

Image

### Reproducer

1. Source ROOT: `source /path/to/root-build/bin/thisroot.sh`
2. Run this Python script:

```python
import os
import tempfile
import numpy as np
import ROOT
from keras import layers, models

model = models.Sequential([
    layers.Input(shape=(2, 5, 5)),
    layers.Conv2D(4, (3, 3), padding='same', strides=(2, 2),
                 data_format='channels_first', activation='relu')
])
model.compile(optimizer='adam', loss='mean_squared_error')
x = np.random.rand(2, 2, 5, 5).astype('float32')
y = np.random.rand(2, 4, 3, 3).astype('float32')
model.fit(x, y, epochs=1, verbose=0)

with tempfile.TemporaryDirectory() as tmpdir:
    model_path = os.path.join(tmpdir, "conv2d_channels_first_same.keras")
    model.save(model_path)
    rmodel = ROOT.TMVA.Experimental.SOFIE.PyKeras.Parse(model_path, batch_size=1)
    hxx_path = os.path.join(tmpdir, "conv2d_channels_first_same.hxx")
    rmodel.Generate()
    rmodel.OutputGenerated(hxx_path)
    ROOT.gInterpreter.Declare('#include "' + hxx_path + '"')
    sess = getattr(ROOT, "TMVA_SOFIE_conv2d_channels_first_same").Session(hxx_path.replace(".hxx", ".dat"))
    km = models.load_model(model_path)
    inp = np.ones((1, 2, 5, 5), dtype='float32')
    sofie_out = np.asarray(sess.infer(inp)).flatten()
    keras_out = np.asarray(km(inp)).flatten()
    print("SOFIE size:", sofie_out.size)
    print("Keras size:", keras_out.size)
```
3. Without fix: SOFIE outputs 24 elements, Keras outputs 36. They should be same.

4. Run: `python script.py` (with ROOT and Keras in path)

Image

### ROOT version

ROOT 6.39.01 

### Installation method

Built from source

### Operating system

Linux (Ubuntu 24.04)

### Additional context

Fix is in `bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_sofie/_parser/_keras/layers/conv.py`. For same padding with channels_first, use `fInputShape[2]` and `fInputShape[3]` for inputHeight and inputWidth instead of `fInputShape[1]` and `fInputShape[2]`.

### PR DESCRIPTION
**Pull request by markknoffler**
Keras Conv2D with `data_format='channels_first'` has shape (N, C, H, W), so H and W are at indices 2 and 3. The parser was always using indices 1 and 2 (correct only for `channels_last` / NHWC) therefore causing wrong padding and output shape mismatches.

Fixes #21661

### Changes
In `bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_sofie/_parser/_keras/layers/conv.py`, the same padding calculation now selects `inputHeight` and `inputWidth` from the correct shape indices depending on `channels_last` vs `channels_first`:
- `channels_last` (default): indices 1 and 2
- `channels_first`: indices 2 and 3

### Checklist
- [x] tested changes locally
- [ ] updated the docs (if necessary)

---

Screenshot 2026-03-23 at 10 18 23 PM

@lmoneta @sanjibansg requesting review.

### CODE REVIEW

**Review by sanjibansg — CHANGES_REQUESTED** [review]
Hi @markknoffler,
Thanks for the PR. This looks good to me. Can you also add a test to verify this patch?

**Comment by markknoffler** [comment]
> Hi @markknoffler, Thanks for the PR. This looks good to me. Can you also add a test to verify this patch?

Respected Sir @sanjibansg,

I have added the requested test. I ran it on both versions: the broken code and the fixed code using separate ROOT builds for each case (before and after the fix).

I have also attached two screenshots showing the run results for both cases.

Kindly let me know if you would like any changes in the test location or format.

With regards,
Samreedh Bhuyan

Screenshot 2026-03-23 at 3 56 18 AM
Screenshot 2026-03-23 at 10 18 23 PM

**Comment by dpiparo** [comment]
The test does not seem to follow the standard way in which tests are handled in ROOT.

**Comment by sanjibansg** [comment]
> The test does not seem to follow the standard way in which tests are handled in ROOT.

I agree with @dpiparo, the tests for the pythonized methods of SOFIE should go into the `bindings` directory, like we have for others: 
1. https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/test/sofie_keras_parser.py
2. https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/test/sofie_gnn.py

**Comment by markknoffler** [comment]
> The test does not seem to follow the standard way in which tests are handled in ROOT.



> > The test does not seem to follow the standard way in which tests are handled in ROOT.
> 
> I agree with @dpiparo, the tests for the pythonized methods of SOFIE should go into the `bindings` directory, like we have for others:
> 
> 1. https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/test/sofie_keras_parser.py
> 2. https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/test/sofie_gnn.py

Respected @dpiparo and @sanjibansg,

Thank you for the feedback and for pointing to the standard location for these tests. I apologise for the earlier approach which did not follow the usual ROOT layout for pythonised SOFIE tests.

I also apologise for delay in updating the PR which was largely due to time spent debugging my local runs since with TensorFlow on CPU and no visible GPU, the helper that gates channels_first models was dropping those cases. So the new test never executed and it was not obvious that the harness was skipping them. I have added an explicit opt-in (ROOT_TMVA_SOFIE_KERAS_CPU_CHANNELS_FIRST=1) so CPU-only runs can still exercise these tests when needed, without changing default behaviour for normal CI.

I have moved the test into the bindings test following the same pattern as the existing Keras parser tests (functional and sequential generators). The new case exercises Conv2D with data_format=channels_first, padding='same', and strides=(2, 2); it fails on the unfixed parser (SOFIE vs Keras output shape mismatch) and passes once the conv.py padding fix is applied.

Going forward I will place pythonised SOFIE tests under bindings/pyroot/pythonizations/test/ and align with the existing patterns for the work that I will submit over the coming months.

With regards,
Samreedh Bhuyan

Before the fix:

Screenshot 2026-03-29 at 12 45 45 AM

after the fix:

Screenshot 2026-03-29 at 2 25 01 AM

**Review by sanjibansg — APPROVED** [review]
LGTM!
Thanks for this implementation. 
@guitargeek @lmoneta can you give this a look please, thanks!

**Review by lmoneta — APPROVED** [review]
LGTM! Thank you for the fix!
